### ISSUE TITLE
Incorrect Jitted Code For a SHL Instruction Resulting From a MEM_WRITE

### ISSUE THREAD

**Issue by trustednstaller** [unicorn-engine/unicorn#1717]
A colleague and I ran across this issue when noticing that the `shl` instruction below neglects to set the `CF`, when it is supposed to.  Upon digging into why this was the case, it turns out that the JIT'd code for this translation block is completely off and comes as a result of a `MEM_WRITE` hook being called when lifting the `shl` to its `TCGOp` representation that is the initial trigger that begins in generating the wrong JIT'd code, and results in the `livenass_pass1` handler removing the necessary, remaining components of the `shl` handler.

<pre><font size="4"><mark class="yellow"><font color=black><b>0045849F 52                        push    edx
004584A0 D3 A4 0C F5 E3 5C 85      shl     dword ptr [esp+ecx-7AA31C0Bh], cl
004584A7 0F 83 80 60 13 00         jnb     loc_58E52D</b></font></mark>
004584AD 87 AC 0C F9 E3 5C 85      xchg    ebp, [esp+ecx+4+var_7AA31C0B]
004584B4 8B 8C 0C E1 E3 5C 85      mov     ecx, [esp+ecx+4+var_7AA31C23]
004584BB 66 C1 64 24 20 CA         shl     word ptr [esp+4+arg_18], 0CAh
004584C1 8B 54 24 20               mov     edx, [esp+4+arg_18]
004584C5 0F 86 FE 10 10 00         jbe     loc_5595C9<br></font></pre>

The resulting `TranslationBlock` is from the `push` up to the `jnb`.  The `shl` is correctly lifted to its IR representation, including the hooks inserted via `gen_jmp_im`.  All of this comes out of `gen_shift_rm_T1`.  However, upon the `liveness_pass1` optimization pass that eventually follows it, the necessary code is "optimized away", resulting in the incorrect JIT'd code, thus resulting into an incorrect interpretation of the flags for this instruction.

![image](https://user-images.githubusercontent.com/107051887/192526524-f8d13f86-6e7a-4783-b346-f6bfc9f52360.png)

Here is the exact location where the hook that gets inserted starts the trouble:

![image](https://user-images.githubusercontent.com/107051887/192527636-aaa94dc8-8dbf-419b-962b-4798ad375e84.png)

The "fix" in our particular case, was simply removing this `MEM_WRITE` hook, which results in avoiding the `gen_jmp_im` being generated and acted upon, but this is a scary fix :)  Will work on putting exact steps for repro, as this is part of a larger emulation we're doing, but I hope it can already begin pinpointing the exact root cause of the issue.

**Update**

Huge kudos to my colleague here w/ the necessary assist.  The resulting code incorrectly generates the same issue w/ the wrong implementation of the `shl` instruction.

```
[0x1968000053c]  : and         ebx,1Fh                    ; missing the sub_i64
[0x1968000053f]  : shlx        r13,r13,rbx                ; missing the subsequent shlx
[0x19680000544]  : mov         qword ptr [rbp+80h],1021h  ; jmp_hook no no no
```

```python
import unicorn
import binascii
from unicorn.x86_const import *
#from capstone import *

def code_hook(mu, access, address, size, value, user_data):
    return True
    
uc = unicorn.Uc(unicorn.UC_ARCH_X86, unicorn.UC_MODE_32)
uc.hook_add(unicorn.UC_HOOK_MEM_READ | unicorn.UC_HOOK_MEM_WRITE, code_hook)

uc.mem_map(0xFFFF0000, 0x2000)
uc.reg_write(UC_X86_REG_ESP, 0xFFFF1000)
uc.reg_write(UC_X86_REG_EBP, 0xFFFF1000)

'''
0x1000:  push    ecx
0x1001:  mov     ecx, 0x1ea53211
0x1006:  lea     ecx, [ecx + ecx*8 + 0x718e20eb]
0x100d:  or      cl, 0xd0
0x1010:  lea     ebp, [ebp + ecx - 0x5778699]
0x1017:  not     ecx
0x1019:  or      dword ptr [esp + ecx - 0x7aa31c0f], ecx
0x1020:  push    edx
0x1021:  shl     dword ptr [esp + ecx - 0x7aa31c0b], cl
0x1028:  jae     0x136087
0x102e:  ret
'''
code = binascii.unhexlify('51b91132a51e8d8cc9eb208e7180c9d08dac0d677988faf7d1098c0cf1e35c8552d3a40cf5e35c850f8380601300C3')
addr = 0x1000

uc.mem_map(0, 0x10000)
uc.mem_write(addr, code)
uc.emu_start(addr, addr + len(code), count=10)
```

The x64 implementation also suffers from the same issue:

```python
import unicorn
import binascii
from unicorn.x86_const import *
#from capstone import *

def code_hook(mu, access, address, size, value, user_data):
    return True
    
'''def debug_print(mu, address, size, user_data):
    md = Cs(CS_ARCH_X86, CS_MODE_64)
    temp = mu.mem_read(address, size)
    for (addr, size, mnemonic, op_str) in md.disasm_lite(temp, 1):
        print("0x%x:\t%s\t%s" %(address, mnemonic, op_str))'''
    
uc = unicorn.Uc(unicorn.UC_ARCH_X86, unicorn.UC_MODE_64)
#uc.hook_add(unicorn.UC_HOOK_MEM_READ | unicorn.UC_HOOK_MEM_WRITE, code_hook) #jmp taken no output
uc.hook_add(unicorn.UC_HOOK_MEM_READ | unicorn.UC_HOOK_MEM_WRITE, code_hook) #jmp not taken crashing resulting in different output:
#  Traceback (most recent call last):
#    File ".\stub_64.py", line 35, in <module>
#      uc.emu_start(addr, addr + len(code), count=5)
#    File "C:\Python38\lib\site-packages\unicorn\unicorn.py", line 341, in emu_start
#      raise UcError(status)
#  unicorn.unicorn.UcError: Invalid memory fetch (UC_ERR_FETCH_UNMAPPED)
#uc.hook_add(unicorn.UC_HOOK_CODE, debug_print)

uc.mem_map(0xFFFF0000, 0x2000)
uc.reg_write(UC_X86_REG_RSP, 0xFFFF1000)
uc.reg_write(UC_X86_REG_RBP, 0xFFFF1000)
'''
0x1000: movabs  rcx, 0xffffffffffffffff
0x100a: mov     qword ptr [rsp], rcx
0x100e: shl     qword ptr [rsp], cl
0x1012: jae     0xd
0x1014: ret
'''
code = binascii.unhexlify('48B9FFFFFFFFFFFFFFFF48890C2448D32424730AC3')
addr = 0x1000

uc.mem_map(0, 0x2000)
uc.mem_write(addr, code)
uc.emu_start(addr, addr + len(code), count=5)
```

**Comment by wtdcode** [comment]
Really good catch, on fixing this.

**Comment by wtdcode** [comment]
Fixed in c889258d8eadf268325d31958b5f2c5ea99f73bb
